include/FixedPointConverter.h

```diff
@@ -331,87 +334,108 @@ namespace ChimeraTK {
-    double d_cooked = _inverseFractionalBitsCoefficient * static_cast<double>(cookedValue);
-
-    // Convert into either signed or unsigned int32_t, depending on _isSigned,
-    // so the conversion handles the sign correctly. Store always in a uint32_t,
-    // since this is our raw type. The conversion will properly round, when
-    // needed. Negative overflow exceptions need to be caught for some corner
-    // cases (e.g. number of fractional bits >= number of bits in total).
-    // Positive overflow cannot happen due to the range check above (the negative
-    // branch has one more possible value).
-    int32_t raw;
-    try {
-      if(_isSigned) {
-        typedef boost::numeric::converter<int32_t, double, boost::numeric::conversion_traits<int32_t, double>,
-            boost::numeric::def_overflow_handler, Round<double>>
-            converter_signed;
-        raw = converter_signed::convert(d_cooked);
+      // handle integer and floating-point types differently
+      if(std::numeric_limits<UserType>::is_integer && _fractionalBits == 0) {
+        // extract the sign and leave the positive number
+        bool isNegative = isNegativeUserType(cookedValue);
+        if(isNegative && !_isSigned) return _minRawValue;
+        if(isNegative) {
+          cookedValue = -(cookedValue + 1); // bit inversion, ~ operator cannot be used as UserType might be double
+        }
+        // cast into raw type
+        auto rawValue = static_cast<RawType>(cookedValue);
+
+        // handle sign
+        if(_isSigned && isNegative) {
+          rawValue = ~rawValue;
+        }
+
+        // return with bit mask applied
+        return rawValue & static_cast<RawType>(_usedBitsMask);
       }
-      else {
-        typedef boost::numeric::converter<uint32_t, double, boost::numeric::conversion_traits<uint32_t, double>,
-            boost::numeric::def_overflow_handler, Round<double>>
-            converter_unsigned;
-        raw = static_cast<int32_t>(converter_unsigned::convert(d_cooked));
+      // convert into double and scale by fractional bit coefficient
+      double d_cooked = _inverseFractionalBitsCoefficient * static_cast<double>(cookedValue);
+
+      // Convert into either signed or unsigned int32_t, depending on _isSigned,
+      // so the conversion handles the sign correctly. Store always in a uint32_t,
+      // since this is our raw type. The conversion will properly round, when
+      // needed. Negative overflow exceptions need to be caught for some corner
+      // cases (e.g. number of fractional bits >= number of bits in total).
+      // Positive overflow cannot happen due to the range check above (the negative
+      // branch has one more possible value).
+      RawType raw;
+      try {
+        if(_isSigned) {
+          using converter_signed = boost::numeric::converter<int32_t, double,
+              boost::numeric::conversion_traits<int32_t, double>, boost::numeric::def_overflow_handler, Round<double>>;
```

Comment:
This one has to convert to, argh., the signed type with the same number of bits as the RawType.